Revise hardware recommendations (#896)
rancher/rancher-docs: docs/reference-guides/best-practices/rancher-server/tuning-and-best-practices-for-rancher-at-scale.md

---
title: Tuning and Best Practices for Rancher at Scale
---

<head>
  <link rel="canonical" href="https://ranchermanager.docs.rancher.com/reference-guides/best-practices/rancher-server/tuning-and-best-practices-for-rancher-at-scale"/>
</head>

> This guide describes the best practices and tuning approaches to scale Rancher setups, and the associated challenges with doing so. As systems grow performance will naturally reduce, but there are steps that can minimize the load put on Rancher, and optimize Rancher's ability to manage larger infrastructures.

Reviewer: Suggestion: This guide describes the best practices and tuning approaches to scale Rancher setups and the associated challenges with doing so. As systems grow, performance will naturally reduce, but there are steps that can minimize the load put on Rancher and optimize Rancher's ability to manage larger infrastructures.

Author: Addressed as of 01dc68f1cbc

* Always scale up gradually, and monitor and observe any changes in behavior while doing do. It is usually easier to resolve performance problems as soon as they surface, before other problems obscure the root cause.

* Reduce network latency between the upstream Rancher cluster and downstream clusters to the extent possible. Note that latency is, among other factors, a function of geographic distance - if you require clusters or nodes spread across the world, consider multiple Rancher installations.

## Minimizing Load on the Upstream Cluster

When scaling up Rancher, one typical bottleneck is resource growth in the upstream (local) Kubernetes cluster. The upstream cluster contains information for all downstream clusters. Many operations that apply to downstream clusters create new objects in the upstream cluster and require computation from handlers running in the upstream cluster.

### Managing Your Object Counts

> Etcd is the backing database for Kubernetes and for Rancher. The database may eventually encounter limitations to the number of a single Kubernetes resource type it can store. Exact limits vary and depend on a number of factors. However, experience indicates that performance issues frequently arise once a single resource type's object count exceeds 60 thousand. Often that type is `RoleBinding`.

Reviewer: 60,000

Author: Addressed as of d5b649a6ca9

---
title: Tuning and Best Practices for Rancher at Scale
---

<head>
  <link rel="canonical" href="https://ranchermanager.docs.rancher.com/reference-guides/best-practices/rancher-server/tuning-and-best-practices-for-rancher-at-scale"/>
:docs/reference-guides/best-practices/rancher-server/tuning-and-best-practices-for-rancher-at-scale.md
</head>

This guide describes best practices and tuning approaches to scale Rancher setups, and associated challenges with doing so. As systems grow performance will naturally reduce, but there are steps that can be taken to minimize the load put on Rancher, as well as optimize Rancher's ability to manage larger infrastructures.

## General Guidelines on Optimizing Rancher's Performance

> * Keep Rancher up to date with patch releases. Performance improvements and bug fixes are made continuously, and the latest release incorporates the largest set ofperformance related development, experience and feedback from many users.

Author: I like the new wording, thanks!

Just out of curiosity: is the use of "we" and "you" generally considered OK for technical documentation? I think I remember some projects ago there was a recommendation to avoid use of such pronouns in favor of passive forms. As a non-native speaker I of course do not have an opinion on my own - those nuances are way too subtle for my sensitivity to even detect 😉

> * Keep Rancher up to date with patch releases. Performance improvements and bug fixes are made continuously, and the latest release incorporates the largest set ofperformance related development, experience and feedback from many users.

Reviewer 2: > @moio 
> I like the new wording, thanks!
> 
> Just out of curiosity: is the use of "we" and "you" generally considered OK for technical documentation? I think I remember some projects ago there was a recommendation to avoid use of such pronouns in favor of passive forms. As a non-native speaker I of course do not have an opinion on my own - those nuances are way too subtle for my sensitivity to even detect 😉

Hi, Rancher docs are a bit different as we actually use the MS style guide instead of the SUSE guide. This is a holdover from when Rancher was a separate company. The MS guide advises against "we" in most cases, while it encourages "you" in place of referring to the reader in the third person (as in "users should..." instead of "you should..."). It also strongly discourages passive voice. Contractions vs no contractions is probably the most obvious different from the SUSE style guide. The MS guide favors contractions, while the SUSE guide does't.

We did consider, as of late, switching to the SUSE guide, but decided against it for the time being. We already have an established docset with hundreds of pages, so standardizing those pages around another style guide would be a big task.

> * Keep Rancher up to date with patch releases. Performance improvements and bug fixes are made continuously, and the latest release incorporates the largest set ofperformance related development, experience and feedback from many users.

Author: That makes total sense, thanks for sharing this background! Will try to keep it in mind for the future.

* Experiment to see if creating new projects or creating new clusters manifests in fewer `RoleBindings` for your specific use case.

### RoleBinding count estimation

Predicting exactly the number of `RoleBindings` a given configuration will create depends on many factors and is complicated to calculate. However, it is possible to give a first estimation according to considerations below:
* As a minimum estimation consider the formula `32C + U + 2UaC + 8P + 5Pa`, where `C` is the cluster count, `U` is the user count, `Ua` is the average count of users with a membership on a cluster, `P` is the project count, and `Pa` is the average number of users with a membership on a project
* The Restricted Admin role follows a different formula, as every user with Restricted Admin role will result in at least `7C + 2P + 2` additional `RoleBindings`
* The number of `RoleBindings` generally increases linearly with cluster count, project count, and user count

### Using New Apps Over Legacy Apps

> There are two app Kubernetes resources that Rancher uses: `apps.projects.cattle.io` and `apps.cattle.cattle.io`. Legacy apps, `apps.projects.cattle.io`, were first introduced with the former UI (Cluster Manager) and are now outdated. New apps, `apps.catalog.cattle.io`, are found in the current UI (Cluster Explorer) for their respective cluster. New apps are preferable because their data resides in downstream clusters, freeing up resources in the local cluster.

Author: Just for my information: what's the current rule for sentences that begin with words that should not be capitalized?

Eg. type names (`apps.cattle.cattle.io`), project names ([etcd](https://etcd.io/docs/v3.4/faq/#what-does-the-etcd-warning-snapshotting-is-taking-more-than-x-seconds-to-finish--mean)), command names...

> There are two app Kubernetes resources that Rancher uses: `apps.projects.cattle.io` and `apps.cattle.cattle.io`. Legacy apps, `apps.projects.cattle.io`, were first introduced with the former UI (Cluster Manager) and are now outdated. New apps, `apps.catalog.cattle.io`, are found in the current UI (Cluster Explorer) for their respective cluster. New apps are preferable because their data resides in downstream clusters, freeing up resources in the local cluster.

Reviewer 2: My instinct was to capitalize, but checking the style guide, it's actually advised to revise so that the always-lowercase word doesn't appear at the start of a sentence

It is recommended to remove any remaining legacy apps that appear in the Cluster Manager, replacing them with apps in the Cluster Explorer for their target cluster if necessary and creating any future apps in the cluster's Cluster Explorer only.

### Using the Authorized Cluster Endpoint (ACE)
An [Authorized Cluster Endpoint](../../../reference-guides/rancher-manager-architecture/communicating-with-downstream-user-clusters#4-authorized-cluster-endpoint) option exist to access the Kubernetes API of Rancher provisioned RKE1, RKE2, and K3s clusters. When enabled this adds a context to generated kubeconfig files generated for the cluster that uses a direct endpoint to the cluster, thereby bypassing Rancher. That reduces load on Rancher for use cases where unmediated API access is acceptable or preferable.

> Note that, in order for `kubeconfig` to take advantage of ACE, users need to issue the `kubectl use-context <ace context name>` command in order to start using it.

Author: Well, that instruction is already in https://ranchermanager.docs.rancher.com/reference-guides/best-practices/rancher-server/tips-for-scaling-rancher#using-the-authorized-cluster-endpoint-ace, which is where it should really be IMO.

I am not sure why this paragraph's original author wanted to stress the point here, I can imagine because of reports of users not getting the expected benefits because context setting was forgotten.

I am personally fine replacing this with a link to the main ACE page if you feel like it does not belong here.

> Note that, in order for `kubeconfig` to take advantage of ACE, users need to issue the `kubectl use-context <ace context name>` command in order to start using it.

Author: Addressed as of 2392110de58

> ### Experimental: Option to Reduce Event Handler Executions

Author: It used to be, but it's not experimental any longer as several users have returned with positive reports (and the fleet project uses the equivalent by default since several versions).

In a future Rancher version, this option will also be made default (and this documentation will be updated accordingly).

> ### Experimental: Option to Reduce Event Handler Executions

Reviewer 2: Maybe edit to "Optional: Reduce Event Handler Executions" so we avoid the "experimental" term entirely

> ### Experimental: Option to Reduce Event Handler Executions

Author: FMPOV it is not any more or less optional than the other steps in this guide - "Option" referred to the environment variable more than an optional step in the intention of the original text.

I changed it to just "Reducing Event Handler Executions" in 6ea7053b233 to follow the same pattern of previous paragraphs, let me know if that is not acceptable from your perspective.

The value for the environment variable can be a comma separated list of the following options. The values refer to types of controllers (the structures that contain and run handlers) and their handlers. Adding the controller types to the variable will disable that set of controllers from running their handlers as part of cache resyncing.

* `mgmt` refers to management controllers which only run on one Rancher node.
* `user` refers to user controllers which run for every cluster. Some of these are ran on the same node as management controllers, while other run in the downstream cluster. This will option targets the former.
* `scaled` refers to scaled controllers which run on every Rancher node. This is not recommended to be set due to the critical functionality the scaled handlers are responsible for.

> In short, if you notice CPU usage peaks every 15 hours, add the CATTLE_SYNC_ONLY_CHANGED_OBJECTS environment variable to your rancher deployment with the value `mgmt,user`.

Author: Well, the Rancher deployment only specifies one container, which only has one place where environment variables can be specified.

If we wanted to guide inexperienced users better we could say something like:

```suggestion
In short, if you notice CPU usage peaks every 15 hours, add the `CATTLE_SYNC_ONLY_CHANGED_OBJECTS` environment variable to your Rancher deployment (in the `spec.containers.env` list) with the value `mgmt,user`
```

## Optimizations Outside of Rancher

> Important influencing factors in Rancher performance are its underlying cluster's own performance and its configuration. The local cluster, if misconfigured, can indeed introduce a bottleneck Rancher software has no chance to resolve.

Reviewer 2: ```suggestion
Important influencing factors are the underlying cluster's own performance and configuration. The upstream cluster, if misconfigured, can introduce a bottleneck Rancher software has no chance to resolve.
```

> ### Manage local cluster nodes directly, use RKE2 as the Kubernetes distribution of choice

Reviewer 2: ```suggestion
### Manage Upstream Cluster Nodes Directly with RKE2

```

> As Rancher can be particularly demanding on the local cluster, especially in large scale scenarios, it is recommended to have full control of its configuration and its nodes. For example, when Rancher nodes experience high resource usage, standard Linux troubleshooting techniques and tools are recommended to identify whether Rancher, Kubernetes components, or OS components are the root cause of the excess resource consumption.

Reviewer 2: ```suggestion
As Rancher can be very demanding on the upstream cluster, especially at scale, you should have full administrative control of the cluster's configuration and nodes. To identify the root cause of excess resource consumption, use standard Linux troubleshooting techniques and tools. This can aid in distinguishing between whether Rancher, Kubernetes, or operating system components are causing issues. 
```

> Consequently, although managed Kubernetes services make it easier to deploy and run Kubernetes clusters, they are discouraged for Rancher's local cluster in high scale scenarios, because they typically limit control on configuration and insights on individual nodes and services.

Reviewer 2: ```suggestion
Although managed Kubernetes services make it easier to deploy and run Kubernetes clusters, they are discouraged for the upstream cluster in high scale scenarios. Managed Kubernetes services typically limit access to configuration and insights on individual nodes and services.
```

> When choosing a Kubernetes distribution, it is recommended to use RKE2 for all Rancher large scale use cases.

Reviewer 2: ```suggestion
Use RKE2 for large scale use cases.
```

> ### Keeping Kubernetes Versions Up to Date

Reviewer 2: ```suggestion
### Keeping Kubernetes Versions Up to Date

```

> Similar to Rancher versions, it is recommended to keep the local Kubernetes cluster up to date. That will ensure that your cluster contains any available performance enhancements and bug fixes.

Reviewer 2: ```suggestion
You should keep the local Kubernetes cluster up to date. This will ensure that your cluster has all available performance enhancements and bug fixes.
```

Reviewer 2: This definitely needs a link to how to update your Kubernetes version. I can look up the proper page.

> ### Optimizing etcd

Reviewer 2: ```suggestion
### Optimizing etcd

```

> etcd is the backing database for Kubernetes and for Rancher, therefore it plays a very important role in Rancher performance.

Reviewer 2: ```suggestion
Etcd is the backend database for Kubernetes and for Rancher. It plays a very important role in Rancher performance.
```

> The two main bottlenecks to [etcd performance](https://etcd.io/docs/v3.4/op-guide/performance/) are disk speed and network speed. It is thus recommended that etcd runs on dedicated nodes with SSDs with high IOPS and a fast network setup. For more information regarding etcd performance see [Slow etcd performance (performance testing and optimization)](https://www.suse.com/support/kb/doc/?id=000020100) and [Tuning etcd for Large Installations](../../../how-to-guides/advanced-user-guides/tune-etcd-for-large-installs). Information on disks can also be found in the [Installation Requirements](../../../pages-for-subheaders/installation-requirements#disks) page.

Reviewer 2: ```suggestion
The two main bottlenecks to [etcd performance](https://etcd.io/docs/v3.4/op-guide/performance/) are disk and network speed. Etcd should run on dedicated nodes with a fast network setup and with SSDs that have high input/output operations per second (IOPS). For more information regarding etcd performance, see [Slow etcd performance (performance testing and optimization)](https://www.suse.com/support/kb/doc/?id=000020100) and [Tuning etcd for Large Installations](../../../how-to-guides/advanced-user-guides/tune-etcd-for-large-installs). Information on disks can also be found in the [Installation Requirements](../../../pages-for-subheaders/installation-requirements#disks).
```

> As adding more nodes in an etcd cluster will make operations slower, for best performance it is recommended to run etcd on exactly 3 nodes. This may be counter-intuitive to common scaling approaches, and it is due to etcd's [replication mechanisms](https://etcd.io/docs/v3.5/faq/#what-is-maximum-cluster-size).

Reviewer 2: ```suggestion
It's best to run etcd on exactly three nodes, as adding more nodes will reduce operation speed. This may be counter-intuitive to common scaling approaches, but it's due to etcd's [replication mechanisms](https://etcd.io/docs/v3.5/faq/#what-is-maximum-cluster-size).
```

> etcd performance will also be negatively affected by network latency between nodes as that will slow down network communication, so it is recommended that etcd nodes are all colocated together with Rancher nodes.

Reviewer 2: ```suggestion
Etcd performance will also be negatively affected by network latency between nodes as that will slow down network communication. Etcd nodes should be located together with Rancher nodes.
```

* If you use [external authentication](../../../pages-for-subheaders/authentication-config), use groups to assign roles.
* Only add users to clusters and projects when necessary.
* Remove clusters and projects when they are no longer needed.
* Only use custom roles if necessary.
* Use as few rules as possible in custom roles.
* Consider whether adding a role to a user is redundant.
* Consider using less, but more powerful, clusters.
* Kubernetes permissions are always "additive" (allow-list) rather than "subtractive" (deny-list). Try to minimize configurations that gives access to all but one aspect of a cluster, project, or namespace, as that will result in the creation of a high number of `RoleBinding` objects.
* Experiment to see if creating new projects or clusters manifests in fewer `RoleBindings` for your specific use case.

### RoleBinding Count Estimation

Predicting how many `RoleBinding` objects a given configuration will create is complicated. However, the following considerations can offer a rough estimate:

> * For a minimum estimate, use the formula `32C + U + 2UaC + 8P + 5Pa`, where `C` is the total number of clusters, `U` is the total number of users, `Ua` is the average number of users with a membership on a cluster, `P` is the total number of projects, and `Pa` is the average number of users with a membership on a project.

Reviewer 3: ```suggestion
* For a minimum estimate, use the formula `32C + U + 2UaC + 8P + 5Pa`. 
  * `C` is the total number of clusters.
  * `U` is the total number of users.
  * `Ua` is the average number of users with a membership on a cluster.
  * `P` is the total number of projects.
  * `Pa` is the average number of users with a membership on a project.
```

Spreading the definitions over multiple lines for readability, similar what line 64-66 has already done,

You should remove any remaining legacy apps that appear in the Cluster Manager UI, and replace them with apps in the Cluster Explorer UI. Create any new apps only in the Cluster Explorer UI.

### Using the Authorized Cluster Endpoint (ACE)

An [Authorized Cluster Endpoint](../../../reference-guides/rancher-manager-architecture/communicating-with-downstream-user-clusters#4-authorized-cluster-endpoint) (ACE) provides access to the Kubernetes API of Rancher-provisioned RKE, RKE2, and K3s clusters. When enabled, the ACE adds a context to kubeconfig files generated for the cluster. The context uses a direct endpoint to the cluster, thereby bypassing Rancher. This reduces load on Rancher for cases where unmediated API access is acceptable or preferable. See [Authorized Cluster Endpoint](../../../reference-guides/rancher-manager-architecture/communicating-with-downstream-user-clusters#4-authorized-cluster-endpoint) for more information and configuration instructions.

> ### Reducing Event Handler Executions

Reviewer 3: ```suggestion
### Reducing Event Handler Executions

```
